packages/transaction-controller/src/utils.ts

```diff
@@ -226,8 +241,122 @@ export function transactionMatchesNetwork(
+ * It will never return TRANSACTION_TYPE_CANCEL or TRANSACTION_TYPE_RETRY as these
+ * represent specific events that we control from the extension and are added manually
+ * at transaction creation.
+ *
+ * @param txParams - Parameters for the transaction.
+ * @param ethQuery - EthQuery instance.
+ * @returns A object with the transaction type and the contract code response in Hex.
+ */
+export async function determineTransactionType(
+  txParams: Transaction,
+  ethQuery: EthQuery,
+): Promise<InferTransactionTypeResult> {
+  let result;
+  let contractCode;
+  const { data, to } = txParams;
+  const name = parseStandardTokenTransactionData(data)?.name;
+
+  if (data && !to) {
+    result = TransactionType.deployContract;
+  } else {
+    const { contractCode: resultCode, isContractAddress } =
+      await readAddressAsContract(ethQuery, to);
+
+    contractCode = resultCode;
+
+    if (isContractAddress) {
+      const hasValue = txParams.value && Number(txParams.value) !== 0;
+
+      const tokenMethodName = [
+        TransactionType.tokenMethodApprove,
+        TransactionType.tokenMethodSetApprovalForAll,
+        TransactionType.tokenMethodTransfer,
+        TransactionType.tokenMethodTransferFrom,
+        TransactionType.tokenMethodSafeTransferFrom,
+      ].find((methodName) => methodName.toLowerCase() === name?.toLowerCase());
+
+      result =
+        data && tokenMethodName && !hasValue
+          ? tokenMethodName
+          : TransactionType.contractInteraction;
+    } else {
+      result = TransactionType.simpleSend;
+    }
+  }
+
+  return { type: result, getCodeResponse: contractCode };
+}
+
+/**
+ * Attempts to decode transaction data using ABIs for three different token standards: ERC20, ERC721, ERC1155.
+ * The data will decode correctly if the transaction is an interaction with a contract that matches one of these
+ * contract standards
+ *
+ * @param data - Encoded transaction data.
+ * @returns A representation of an ethereum contract call.
+ */
+export function parseStandardTokenTransactionData(
+  data?: string,
+): TransactionDescription | undefined {
+  if (!data) {
+    return undefined;
+  }
+
+  try {
+    return erc20Interface.parseTransaction({ data });
+  } catch {
+    // ignore and next try to parse with erc721 ABI
+  }
+
+  try {
+    return erc721Interface.parseTransaction({ data });
+  } catch {
+    // ignore and next try to parse with erc1155 ABI
+  }
+
+  try {
+    return erc1155Interface.parseTransaction({ data });
+  } catch {
+    // ignore and return undefined
+  }
+
+  return undefined;
+}
+
+/**
+ * Reads an Ethereum address and determines if it is a contract address.
+ *
+ * @param ethQuery - The Ethereum query object used to interact with the Ethereum blockchain.
+ * @param address - The Ethereum address.
+ * @returns An object containing the contract code and a boolean indicating if it is a contract address.
+ */
+export async function readAddressAsContract(
```

Comment:
As there is a good amount of code to determine the transaction type, plus the new constants, would it be nicer to modularise this a little and create a `transaction-type.ts` to hold all these new functions?

Reply:
I agree it's a substantial amount of code, pushed the changes to support it.